bindings/rust/s2n-tls/src/testing.rs

```diff
@@ -325,3 +335,209 @@ pub fn poll_tls_pair_result(pair: &mut Pair<Harness, Harness>) -> Result<()> {
+    fn register_connection(
+        mode: enums::Mode,
+        config: &config::Config,
+        send_ctx: &Pin<Box<LocalDataBuffer>>,
+        recv_ctx: &Pin<Box<LocalDataBuffer>>,
+    ) -> Result<connection::Connection, error::Error> {
+        let mut conn = connection::Connection::new(mode);
+        conn.set_config(config.clone())?
+            .set_blinding(Blinding::SelfService)?
+            .set_send_callback(Some(Self::send_cb))?
+            .set_receive_callback(Some(Self::recv_cb))?;
+        unsafe {
+            // cast 1 : send_ctx as &LocalDataBuffer -> get a plain reference to underlying LocalDataBuffer
+            //
+            // cast 2: &LocalDataBuffer as *const LocalDataBuffer -> cast the reference to a pointer
+            //     SAFETY: the LocalDataBuffer must live as long as the connection does. This can be violated if the
+            //             connections are moved out from the struct.
+            //
+            // cast 3: *const LocalDataBuffer as *mut c_void -> cast into the final *mut c_void required
+            //     SAFETY: serialized access is enforced by the interior RefCell, so it is safe to vend out
+            //             multiple mutable pointers to this item. We ensure this by casting back to an immutable
+            //             reference in the send and recv callbacks
+            conn.set_send_context(
+                send_ctx as &LocalDataBuffer as *const LocalDataBuffer as *mut c_void,
+            )?
+            .set_receive_context(
+                recv_ctx as &LocalDataBuffer as *const LocalDataBuffer as *mut c_void,
+            )?;
+        }
+        Ok(conn)
+    }
+
+    /// perform a TLS handshake between the connections
+    ///
+    /// This method will call `poll_negotiate` on each connection until both return
+    /// Ready(Ok) which indicates a successful handshake, or until one of the connections
+    /// returns Ready(Err) which indicates some fatal error.
+    pub fn handshake(&mut self) -> Result<(), error::Error> {
+        loop {
+            match (self.client.poll_negotiate(), self.server.poll_negotiate()) {
+                // if everything is finished and Ok, return Ok
+                (Poll::Ready(Ok(_)), Poll::Ready(Ok(_))) => return Ok(()),
+                // if there has been an error on the server
+                (_, Poll::Ready(Err(e))) => return Err(e),
+                // or an error on the client, then return the error
+                (Poll::Ready(Err(e)), _) => return Err(e),
+                _ => { /* not ready, poll again */ }
+            }
+        }
+    }
+
+    unsafe extern "C" fn send_cb(context: *mut c_void, data: *const u8, len: u32) -> c_int {
+        let context = &*(context as *const LocalDataBuffer);
+        let data = core::slice::from_raw_parts(data, len as _);
+        context
+            .borrow_mut()
+            .write(data)
+            .unwrap()
+            .try_into()
+            .unwrap()
```

Comment:
How is this ultimately producing the length of the newly added data as a c_int? If there was already some data in the send buffer would this indicate that more than `len` data was sent?

Reply:
`write(data)` returns the number of bytes that were successfully written. I shifted some of my naming to try and make that a bit clearer.